File: search-a-2d-matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        ios_base::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);
        int r=matrix.size();
        int c=matrix[0].size();
        int i=0,j=c*r-1;
        while(i<=j){
            int m=i+(j-i)/2;
            int x=matrix[m/c][m%c];
            if(x<target)
                i=m+1;
            else if(x>target)
                j=m-1;
            else
                return 1;
        }
        return 0;
    }
};

File: search-a-2d-matrix/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        r=len(matrix)
        c=len(matrix[0])
        i=0
        j=c*r-1
        while i<=j:
            m=i+(j-i)//2
            x=matrix[m//c][m%c]
            if x<target:
                i=m+1
            elif x>target:
                j=m-1
            else:
                return True
        return False